feat: Builder pre claim (#2253)
contentauth/c2pa-rs: docs/builder-pre-claim.md

# Builder pre-claim: what changed and why

## Summary

Previously, `Builder` accumulated everything into `ManifestDefinition` (a JSON-serializable struct) and only built the actual `Claim` once, inside `to_claim()`, at signing time. This branch adds a `Claim` up front (`Builder::pre_claim: Option<Claim>`) and lets several methods write directly into it as they're called, instead of staging data in `definition` for later translation.

> `to_claim()` still runs the old translation logic for anything left in `definition` (JSON-set title/format, `definition.assertions`, `definition.ingredients`, `definition.thumbnail`, actions, etc.) — it clones `pre_claim` (if present) as its starting point, then does its usual pass over `definition`, skipping anything that's already been added directly. Both models coexist in the same `Builder`; nothing is removed.

Reviewer: ```suggestion
`to_claim()` still runs the old translation logic for anything left in `definition` (JSON-set title/format, `definition.assertions`, `definition.ingredients`, `definition.thumbnail`, actions, etc.). It clones `pre_claim` (if present) as its starting point, then does its usual pass over `definition`, skipping anything that's already been added directly. Both models coexist in the same `Builder`.
```
(Shorter sentences are easier to read :) )

## Why

> Some assertions need to reference other assertions by hashed URI before the manifest is signed (this is an increasingly common pattern in the C2PA spec — e.g. an action referencing the ingredient it operated on). Under the old model, an assertion added via `add_assertion` doesn't have a knowable JUMBF URI/hash until `to_claim()` runs, so there was no way to embed such a reference directly; call sites either duplicated logic or referenced things positionally/by id, resolved later.

Reviewer: ```suggestion
Some assertions need to reference other assertions by hashed URI before the manifest is signed (e.g. an action referencing the ingredient it operated on). Under the previous model, an assertion added via `add_assertion` doesn't have a knowable JUMBF URI/hash until `to_claim()` runs, so there was no way to embed such a reference directly; call sites either duplicated logic or referenced things positionally/by id, resolved later.
```

### `Builder::add_embedded_data(&mut self, label: &str, format: &str, stream: &mut impl Read) -> Result<HashedUri>`

Convenience wrapper over `add_assertion_with_ref` for binary data (thumbnails, icons, arbitrary resources): wraps the bytes in an `EmbeddedData` assertion. `set_thumbnail`/`add_resource` now route through this for v2+ claims instead of the `ResourceStore`/`ResourceRef`-by-identifier path.

### `Builder::add_ingredient_with_reader<T>(&mut self, ingredient_assertion: T, reader: &Reader, redactions: Option<Vec<String>>) -> Result<HashedUri>`

Adds an ingredient assertion built from an already-read/validated `Reader`, without constructing a full `Ingredient`. It merges the reader's own manifest data into the pre-claim's ingredient store, fills in `activeManifest`/`claimSignature`/`validationResults` on the assertion, and returns the assertion's `HashedUri`.

`T` accepts, via `TryInto<IngredientAssertion>`:
- an `IngredientAssertion` (owned) or `&IngredientAssertion`
- JSON text (`&str`/`String`) or `serde_json::Value`, matching the assertion's own v3 JSON schema (`dc:title`, `dc:format`, `relationship`, `thumbnail`, etc.)

> `redactions` is a list of JUMBF URIs of assertions to strip from the ingredient's manifest chain as it's merged in — find the URI via the `Reader` (`reader.active_label()` + `jumbf::labels::to_assertion_uri(label, assertion_label)`), then pass it here. The caller is responsible for separately recording a `c2pa.redacted` action documenting why, same as the existing redaction pattern elsewhere in `Builder` — this method only does the mechanical redaction.

Reviewer: With this change, actually we could consider an add_reaction API which takes a URI and adds the action, no?

> `redactions` is a list of JUMBF URIs of assertions to strip from the ingredient's manifest chain as it's merged in — find the URI via the `Reader` (`reader.active_label()` + `jumbf::labels::to_assertion_uri(label, assertion_label)`), then pass it here. The caller is responsible for separately recording a `c2pa.redacted` action documenting why, same as the existing redaction pattern elsewhere in `Builder` — this method only does the mechanical redaction.

Reviewer 2: Why not automatically add the `c2pa.redacted` action? Leaving it to the caller seems error-prone.

Reviewer 2: Also, I'd argue for removing the `Option<...>`. An empty `Vec<>` does no heap allocation, so it's very cheap.

For an ingredient with no manifest data of its own, skip this method: build the `IngredientAssertion` directly and add it with `add_assertion_with_ref`.

### `Action::add_ingredient_ref(self, ingredient: HashedUri) -> Self`

Adds an ingredient reference directly from a `HashedUri` (e.g. the one returned by `add_ingredient_with_reader`), instead of by id resolved later in `to_claim()`. Matches the existing `add_ingredient_id` incremental pattern — call it once per ingredient.

### `IngredientAssertion` is now a public type

> `assertions::Ingredient` (re-exported as `IngredientAssertion`) was previously `pub(crate)` only. It's now public, and gained `Clone`, plus `TryFrom<&str>` / `TryFrom<String>` / `TryFrom<&String>` / `TryFrom<serde_json::Value>` / a `Deserialize` impl for its v3 JSON schema (this re-encodes the input as CBOR and delegates to the existing `from_assertion` decoder, so there's one source of truth for the v3 field set).

Reviewer 2: ```suggestion
`assertions::Ingredient` (re-exported as `IngredientAssertion`) was previously `pub(crate)` only. It's now public, and gained `Clone`, plus `TryFrom<&str>` / `TryFrom<String>` / `TryFrom<&String>` / `TryFrom<serde_json::Value>` / a `Deserialize` impl for its v3 JSON schema. (This re-encodes the input as CBOR and delegates to the existing `from_assertion` decoder, so there's one source of truth for the v3 field set.)
```

The parenthetical comment is a complete sentence, so should be punctuated accordingly.
